### ISSUE TITLE
Definite assignment not detected in conditional access

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#36218]
The following sample program produces an error, even though we might expect that within the "then" block an assignment is known to have occurred.

```cs
public class C
{
    static void M1(C c)
    {
        string s1;
        if (c?.M(out s1) == true)
        {
            Console.WriteLine(s1); // error CS0165: Use of unassigned local variable 's1'
        }
    }
    
    bool M(out string s) { s = ""; return true; }
}
```

Compare to this similar program which does detect a definite assignment.
```cs
public class C
{
    static void M2(C c)
    {
        string s2;
        if (c != null && c.M(out s2))
        {
            Console.WriteLine(s2); // OK
        }
    }
    
    bool M(out string s) { s = ""; return true; }
}
```

**Comment by CyrusNajmabadi** [comment]
I think this would require the compiler to know that `default(bool?)` was `!=` to `true`.  That feels like something it shoudl have knowledge about as `default(Nullable)` should always be non-equal to a non-null struct value.

**Comment by RikkiGibson** [comment]
NullableWalker handles these kinds of scenarios with some code specific to ConditionalAccess: http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/FlowAnalysis/NullableWalker.cs,2171

It might be necessary to generalize this or implement something similar over in DefiniteAssignment to make this scenario work. But any work on this should probably start with studying the definite assignment and conditional access specs to see if the language supports this scenario or if a spec revision is needed.

**Comment by gafter** [comment]
There are no special language-specified definite assignment rules for `== true` that would make the original example work as desired. Roslyn implements the language specification, and therefore this is by design.  If you would like the language specification changed, please submit a specific proposal (i.e. how you propose to change the specification) to csharplang.
